src/c_api.cpp

```diff
@@ -185,8 +185,44 @@ class Booster {
     if (param.count("metric")) {
       Log::Fatal("Cannot change metric during training");
     }
-
+    auto old_config = config_;
     config_.Set(param);
+
+    if (param.count("max_bin") && config_.max_bin != old_config.max_bin) {
```

Comment:
@guolinke Can't `LGBM_DatasetUpdateParamWarning` be called here to avoid code duplication?